include/circt/Dialect/Seq/Seq.td

```diff
@@ -60,4 +60,47 @@ def CompRegOp : SeqOp<"compreg",
   ];
 }
 
+def FirRegOp : SeqOp<"firreg",
+    [NoSideEffect, AllTypesMatch<["input", "data"/*, "resetValue"*/]>,
+     SameVariadicOperandSize,
+     DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>]> {
+      // AllTypesMatch doesn't work with Optional types yet.
+
+  let summary = "Register with preset and sync or async reset";
+  let description = [{
+    See the Seq dialect rationale for a longer description.
+
+    ```
+    %name = seq.firreg %input, %clk [ sym @sym]
+        [ reset (sync|async) %reset, %value ] : type
+    ```
+  }];
```

Comment:
I borrowed this from `compreg`. Optional operands also count as variadic operands, which require some constraints on their size and number. We could use `AttrSizedOperandSegments`, but that actually includes some attributes on the op. `SameVariadicOperandSize` is pretty much a no-op when there is a single variadic operand, without requiring an extra attribute. 

The strategy for presets is to include additional syntax and attributes to specify something along the lines of `seq.firreg ... preset random` or `seq.firreg preset %constant`.

At some point, we might want to re-think sync resets, maybe removing them and adding a mux. However, for now, they're quite useful since we can emit an if statement for them and that really helps readability.